src/lib/tls/tls13_pqc/hybrid_public_key.cpp

```diff
@@ -0,0 +1,365 @@
+   const auto algo_spec = algorithm_specs_for_group(group);
+   std::vector<std::unique_ptr<Private_Key>> private_keys;
+   private_keys.reserve(algo_spec.size());
+   for(const auto& spec : algo_spec) {
+      private_keys.push_back(create_private_key(spec.first, rng, spec.second));
+   }
+   return std::make_unique<Hybrid_KEM_PrivateKey>(std::move(private_keys));
+}
+
+Hybrid_KEM_PrivateKey::Hybrid_KEM_PrivateKey(std::vector<std::unique_ptr<Private_Key>> sks) :
+      Hybrid_KEM_PublicKey(extract_public_keys(sks)) {
+   BOTAN_ARG_CHECK(!sks.empty(), "List of private keys must not be empty");
+   BOTAN_ARG_CHECK(std::all_of(sks.begin(),
+                               sks.end(),
+                               [](const auto& sk) {
+                                  return sk->supports_operation(PublicKeyOperation::KeyEncapsulation) ||
+                                         sk->supports_operation(PublicKeyOperation::KeyAgreement);
+                               }),
+                   "Some provided private key is not compatible with this hybrid wrapper");
+
+   std::transform(
+      sks.begin(), sks.end(), std::back_inserter(m_private_keys), [](auto& key) -> std::unique_ptr<Private_Key> {
+         if(key->supports_operation(PublicKeyOperation::KeyAgreement) &&
+            !key->supports_operation(PublicKeyOperation::KeyEncapsulation)) {
+            auto ka_key = dynamic_cast<PK_Key_Agreement_Key*>(key.get());
+            BOTAN_ASSERT_NONNULL(ka_key);
+            (void)key.release();
+            return std::make_unique<KEX_to_KEM_Adapter_PrivateKey>(std::unique_ptr<PK_Key_Agreement_Key>(ka_key));
+         } else {
+            return std::move(key);
+         }
+      });
+}
+
+secure_vector<uint8_t> Hybrid_KEM_PrivateKey::private_key_bits() const {
+   throw Not_Implemented("Hybrid private keys cannot be serialized");
+}
+
+std::unique_ptr<Public_Key> Hybrid_KEM_PrivateKey::public_key() const {
+   std::vector<std::unique_ptr<Public_Key>> pks;
+   pks.reserve(m_private_keys.size());
+   for(const auto& sk : m_private_keys) {
+      pks.push_back(sk->public_key());
+   }
+   return std::make_unique<Hybrid_KEM_PublicKey>(std::move(pks));
+}
+
+bool Hybrid_KEM_PrivateKey::check_key(RandomNumberGenerator& rng, bool strong) const {
+   return reduce(m_public_keys, true, [&](bool ckr, const auto& key) { return ckr && key->check_key(rng, strong); });
+}
+
+namespace {
+
+class Hybrid_KEM_Decryption final : public PK_Ops::KEM_Decryption_with_KDF {
+   public:
+      Hybrid_KEM_Decryption(const Hybrid_KEM_PrivateKey& key,
+                            RandomNumberGenerator& rng,
+                            const std::string_view kdf,
+                            const std::string_view provider) :
+            PK_Ops::KEM_Decryption_with_KDF(kdf) {
+         for(const auto& private_key : key.private_keys()) {
+            PK_KEM_Encryptor enc(*private_key, "Raw");
```

Comment:
With #3653 it will not be required anymore to create a `PK_KEM_Encryptor` just to get the encapsulated length

Reply:
#3653 is merged now so this can use `dec.encapsulated_key_length`

Reply:
Very nice! I rebased the PR and used the new `encapsulated_key_length` methods in 0988317.